tool/tsh/putty_config_windows.go

```diff
@@ -0,0 +1,309 @@
+	puttySessionName := fmt.Sprintf(`%v%%20(%v)`, hostname, proxyHostname)
+	registryKey := fmt.Sprintf(`%v\%v`, puttyRegistrySessionsKey, puttySessionName)
+	sessionDwords := puttyRegistrySessionDwords{}
+	sessionStrings := puttyRegistrySessionStrings{}
+
+	// if the port passed is 0, this means "use server default" so we override it to 3022
+	if port == 0 {
+		port = puttyDefaultSSHPort
+	}
+
+	sessionDwords = puttyRegistrySessionDwords{
+		Present:        puttyDwordPresent,
+		PortNumber:     port,
+		ProxyPort:      puttyDefaultProxyPort,
+		ProxyMethod:    puttyDwordProxyMethod,
+		ProxyLogToTerm: puttyDwordProxyLogToTerm,
+	}
+
+	sessionStrings = puttyRegistrySessionStrings{
+		Hostname:            hostname,
+		Protocol:            puttyProtocol,
+		ProxyHost:           proxyHostname,
+		ProxyUsername:       login,
+		ProxyPassword:       "",
+		ProxyTelnetCommand:  commandToRun,
+		PublicKeyFile:       ppkFilePath,
+		DetachedCertificate: certificateFilePath,
+		UserName:            login,
+	}
+
+	// now check for and create the individual session key
+	pk, err := getRegistryKey(registryKey)
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+	defer pk.Close()
+
+	// write dwords
+	if ok, err := registryWriteDword(pk, "Present", sessionDwords.Present); !ok {
+		return false, trace.Wrap(err)
+	}
+	if ok, err := registryWriteDword(pk, "PortNumber", fmt.Sprintf("%v", sessionDwords.PortNumber)); !ok {
+		return false, trace.Wrap(err)
+	}
+	if ok, err := registryWriteDword(pk, "ProxyPort", fmt.Sprintf("%v", sessionDwords.ProxyPort)); !ok {
+		return false, trace.Wrap(err)
+	}
+	if ok, err := registryWriteDword(pk, "ProxyMethod", sessionDwords.ProxyMethod); !ok {
+		return false, trace.Wrap(err)
+	}
+	if ok, err := registryWriteDword(pk, "ProxyLogToTerm", sessionDwords.ProxyLogToTerm); !ok {
+		return false, trace.Wrap(err)
+	}
+
+	// write strings
+	if ok, err := registryWriteStrings(pk, "Hostname", sessionStrings.Hostname); !ok {
+		return false, trace.Wrap(err)
+	}
+	if ok, err := registryWriteStrings(pk, "Protocol", sessionStrings.Protocol); !ok {
+		return false, trace.Wrap(err)
+	}
+	if ok, err := registryWriteStrings(pk, "ProxyHost", sessionStrings.ProxyHost); !ok {
+		return false, trace.Wrap(err)
+	}
+	if ok, err := registryWriteStrings(pk, "ProxyUsername", sessionStrings.ProxyUsername); !ok {
+		return false, trace.Wrap(err)
+	}
+	if ok, err := registryWriteStrings(pk, "ProxyTelnetCommand", sessionStrings.ProxyTelnetCommand); !ok {
+		return false, trace.Wrap(err)
+	}
+	if ok, err := registryWriteStrings(pk, "PublicKeyFile", sessionStrings.PublicKeyFile); !ok {
+		return false, trace.Wrap(err)
+	}
+	if ok, err := registryWriteStrings(pk, "DetachedCertificate", sessionStrings.DetachedCertificate); !ok {
+		return false, trace.Wrap(err)
+	}
+	if ok, err := registryWriteStrings(pk, "UserName", sessionStrings.UserName); !ok {
+		return false, trace.Wrap(err)
+	}
```

Comment:
I don't like this at all, but couldn't find a decent way to iterate over structs that didn't involve a lot of usage of `reflect.VisibleFields()` . What's a more idiomatic way to do this in Go?

Reply:
If you really wanted to, you could use reflection and tags, similar to JSON. Might surface something like:

```
type registryValues struct {
    Present  string `registry:"Present,dword"`
    Port     int    `registry:"PortNumber,dword"`
    Hostname string `registry:Hostname,string"`
}
```

Given that this is a relatively isolated change, I'm not convinced it would be any better. This code, while a bit repetitive, is pretty easy to follow.